pages/blog/internals-of-nova-part-2.md

```diff
@@ -0,0 +1,286 @@
+
+Data-oriented design is all about solving or at least mitigating this issue. To do so, we must know our data and use that knowledge to minimize memory usage which always improves cache usage, and we must take into account the CPU caches when desigining our software. The important thing here is that a CPU loads data into cache not byte-by-byte but in cache lines. This is usually 64 bytes at a time, though recent CPUs have bumped this number up to 128 bytes.
+
+```rs
+#[repr(align(64))]
+struct CacheLine([u8; 64]);
+```
+
+When we load a piece of data we load a full cache line into the CPU, no matter how small a part of it we actually want to use. Loading this one cache line from main memory takes 200 nanoseconds. Now, if we only use a single byte (or worse, one bit!) of that data then the vast majority of the data loaded is wasted. What we want is to ensure that as much of the cache line gets used, especially if we are in a tight loop.
+
+Let's take as an example an Array of Arrays in the V8 JavaScript engine.
+
+```ts
+const arr: Data[][] = await Promise.all(names.map(fetchDataForName));
+```
+
+Now say we are interested in how many `Data` entities there are in total: We will loop over the `arr` Array, accessing each Array in it, and tallying up the length:
+
+```ts
+const dataCount = arr.reduce((acc, dataArray) => acc + dataArray.length, 0);
+```
+
+In terms of memory access patterns, a V8 Array looks roughly like this:
+
+```rs
+struct Array(*mut ArrayHeapData);
+
+// Size is 32 bytes in Node, 16 bytes in Chrome with pointer compression
+struct ArrayHeapData {
+    map: *mut Map,
+    elements: *mut FixedArray,
+    length: u32,
+    properties: *mut FixedArray,
+}
+```
+
+When we reduce through `arr`, we must first access the `elemeents`. In this case it gives us roughly something like:
+
+```rs
+struct FixedArray<const N: usize> {
+    length: usize,
+    entries: [*mut ArrayHeapData; N],
+}
+```
+
+From here we can chase each `*mut ArrayHeapData` to get to the `length` data contained in each, which we then load and sum up.
+
+It is fair to assume that most of these `ArrayHeapData` structs are located close to each other: We shall assume that they are actually right next to one another which means that we can fit two of them on one cache line (four in Chrome). What we want out of each `ArrayHeapData` is a single `u32`, that is 4 bytes. This means that we use 8 bytes (16 in Chrome) of every 64-byte cache line. That is a usage rate of 12.5%. In expert circles this is called "bad".
+
+Because we assumed that all the `ArrayHeapData` are right next to each other, the CPU will actually aggressively prefetch data for us during this reduce action and we might find that we only pay the memory latency a few times, and after that we are bound by the memory bandwidth. Hearing this, you might think that we it is actually fine that the cache usage rate is so low: The CPU prefetching handled any possible problems we might have otherwise had.
```

Comment:
```suggestion
Because we assumed that all the `ArrayHeapData` are right next to each other, the CPU will actually aggressively prefetch data for us during the reduce operation and we might find that we only need to pay the increased memory latency of main memory access a few times, after which we are bound by the memory bandwidth. Hearing this, you might think that we it is actually fine that the cache usage rate is so low: The CPU prefetching handled any possible problems we might have otherwise had.
```
**suggestion**: Could use some re-wording/structuring. Also are we talking about the memory bandwidth of the CPU, RAM, L-Caches?

Reply:
This is actually pretty badly sourced. I don't _really_ understand CPUs deeply enough, I'm just kind of guessing/assuming that there are three (four) possible bounds to an algorithm's / process' / computer's performance:
1. CPU bound: We have all the data we need, but just cannot execute computations any faster than this.
2. Memory bandwidth bound: We do not have all the data we need but we know exactly where it all is well enough in advance to start streaming it in before we actually need it. But, we cannot load data from RAM "in a wide enough stream" to saturate our CPU (or the bandwidth is just badly utilised). The entire memory bus is 100% utilised the whole time, it's just not enough.
3. Memory latency bound: We do not have all the data we need and we do not know where it all is before we need it.
4. I/O bound: God damn spinning disk of rust! Spin faster!

... To be continued...

Reply:
So my guess/assumption here is that the very first steps of the algorithm will be effectively memory latency bound. After that, we will be streaming data in via prefetching and processing it up as it comes. But: Our data isn't particularly dense and our computations aren't particularly heavy so we end up not being CPU bound; we can easily do all the computations we need to before we need to make room for the next "step" of the incoming data stream. Most of the time the CPU is idle, waiting for the next data to come in. When it comes in, we perform a few CPU operations and go right back to idling.